### DISCUSSION TITLE
Unhandled Exception in RabbitMQ Message Consumer Thread, was not caught by AppDomain.CurrentDomain.UnhandledException

### DISCUSSION THREAD

**Discussion by Pramod34** [rabbitmq/rabbitmq-dotnet-client#1650] [Q&A] [answered]
Hi,

We have an application which throws an unhandled exception in the message consumer thread, but AppDomain.CurrentDomain.UnhandledException event didn't get triggered. I have attached the some example code below.

Please help us. We would be appreciate if you can schedule a meeting to review this case.

Thanks,

Pramod

Code Example:

```csharp
AppDomain.CurrentDomain.UnhandledException += AppDomain_UnhandledException;

consumer.Received += (model, ea) =>
                {
                    var body = ea.Body.ToArray();
                    var props = ea.BasicProperties;

                    throw new Exception("consumer errror 2");

                };
```

**Comment by michaelklishin** [comment]
@Pramod34 we will not schedule any meetings with users who are neither paying customers nor contributors.

We do not have any relevant information to work with:

 * Client library version
 * .NET version used
 * An explanation as to what you expect `AppDomain.CurrentDomain.UnhandledException` to be thrown

In general, RabbitMQ clients expect applications to handle consumer exceptions because libraries
themselves have no contexts to do much about them. Logging and (potentially) propagating them is the only thing that a library can do.

It may be reasonable to expect un unhandled exception to be propagated as a `AppDomain.CurrentDomain.UnhandledException` but your "report" lacks the details but you had the guts to ask for "a meeting to be scheduled."

**Reply by Pramod34** [reply]
Hi,

I have posted it in my organization internal group for help, so I have requested them for meeting. I have samething posted here. I apologize it.

We are using following versions:
Client lib: 5.2.0
.NET Version 4.5.1

We have library which has all the our RabbitMQ functionality. That library is being used by  multiple applications, some applications are getting random unhandled exceptions. When we get those unhandled exceptions we want application to shutdown/exit. We have AppDomain.CurrentDomain.UnhandledException logic in our library and expected to hit event when exception came and log the exception on log and shutdown. When shutdown happens we have other service running, which continuously watches any shutdowns and restarts the application them automatically.

Can you please help me in solving this issue?  

Thanks,
Pramod

**Reply by michaelklishin** [reply]
@Pramod34 we will not help you with your internal library, that's literally your job. All the decisions in that library are your decisions and your responsibility.

How you decide to monitor unhandled exceptions in your internal library should not drive the design decisions of the client, and even more so when you use a 5.x series (you are almost two **major** releases behind, 7.x is about to be released).

**Comment by lukebakken** [comment, marked as answer]
@Pramod34 -

I created some example code to show how to deal with exceptions in event handlers, like for the `Received` event:

https://github.com/lukebakken/rabbitmq-dotnet-client-1650/blob/main/Program.cs

When you run that console application, and publish a message to the queue, you can see that the `CallbackException` handler is called.

We do not support version 5 of this library, however, this is the code in version 6 that calls callback exception handlers:

https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/6.x/projects/RabbitMQ.Client/client/impl/ModelBase.cs#L400-L416

You can see that the exception is swallowed.

In your case, you should use the `OnConnectionCallbackException` handlers to _SIGNAL_ another part of your code to exit, and then check to see if there should be a top-level exception re-thrown. See my example code.
